File: tests/typesafety/python/dispatch_typed_misuse.py
# expect: fails
# title: A handler that takes the invoice number, rejected by the checker
# misuse-of: dispatch_typed
from collections.abc import Callable
from dataclasses import dataclass
from typing import Literal

type Status = Literal["draft", "sent", "paid"]


@dataclass(frozen=True)
class Invoice:
    number: str
    status: Status


def send(invoice: Invoice) -> Invoice:
    return Invoice(invoice.number, "sent")


def archive(number: str) -> str:
    return f"archived {number}"


HANDLERS: dict[Status, Callable[[Invoice], Invoice]] = {
    "draft": send,
    "sent": archive,
    "paid": send,
}

File: tests/typesafety/typescript/dispatch_typed_misuse.ts
// expect: fails

type Status = "draft" | "sent" | "paid";

type Invoice = { readonly number: string; readonly status: Status };

function send(invoice: Invoice): Invoice {
  return { number: invoice.number, status: "sent" };
}

function archive(invoiceNumber: string): string {
  return `archived ${invoiceNumber}`;
}

const handlers: Record<Status, (invoice: Invoice) => Invoice> = {
  draft: send,
  sent: archive,
  paid: send,
};
